cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,680 @@
+                                      stream);
+
+          read_tasks.emplace_back(std::move(future_read_size));
+        } else {
+          buffer                   = source->host_read(bitset_offset, bitset_size);
+          bloom_filter_data[chunk] = rmm::device_buffer{buffer->data(), buffer->size(), stream};
+        }
+      }
+    });
+
+  // Read task sync function
+  auto sync_fn = [](decltype(read_tasks) read_tasks) {
+    for (auto& task : read_tasks) {
+      task.wait();
+    }
+  };
+
+  std::async(std::launch::async, sync_fn, std::move(read_tasks)).wait();
+}
+
+}  // namespace
+
+std::vector<rmm::device_buffer> aggregate_reader_metadata::read_bloom_filters(
+  host_span<std::unique_ptr<datasource> const> sources,
+  host_span<std::vector<size_type> const> row_group_indices,
+  host_span<int const> column_schemas,
+  size_type total_row_groups,
+  rmm::cuda_stream_view stream) const
+{
+  // Descriptors for all the chunks that make up the selected columns
+  auto const num_input_columns = column_schemas.size();
+  auto const num_chunks        = total_row_groups * num_input_columns;
+
+  // Association between each column chunk and its source
+  std::vector<size_type> chunk_source_map(num_chunks);
+
+  // Keep track of column chunk file offsets
+  std::vector<std::optional<int64_t>> bloom_filter_offsets(num_chunks);
+  std::vector<std::optional<int32_t>> bloom_filter_sizes(num_chunks);
+
+  // Gather all bloom filter offsets and sizes.
+  size_type chunk_count = 0;
+
+  // For all data sources
+  std::for_each(thrust::make_counting_iterator<size_t>(0),
+                thrust::make_counting_iterator(row_group_indices.size()),
+                [&](auto const src_index) {
+                  // Get all row group indices in the data source
+                  auto const& rg_indices = row_group_indices[src_index];
+                  // For all row groups
+                  std::for_each(rg_indices.cbegin(), rg_indices.cend(), [&](auto const rg_index) {
+                    // For all column chunks
+                    std::for_each(
+                      column_schemas.begin(), column_schemas.end(), [&](auto const schema_idx) {
+                        auto& col_meta = get_column_metadata(rg_index, src_index, schema_idx);
+
+                        // Get bloom filter offsets and sizes
+                        bloom_filter_offsets[chunk_count] = col_meta.bloom_filter_offset;
+                        bloom_filter_sizes[chunk_count]   = col_meta.bloom_filter_length;
+
+                        // Map each column chunk to its source index
+                        chunk_source_map[chunk_count] = src_index;
+                        chunk_count++;
+                      });
+                  });
+                });
+
+  // Do we have any bloom filters
+  if (std::any_of(bloom_filter_offsets.cbegin(),
+                  bloom_filter_offsets.cend(),
+                  [](auto const offset) { return offset.has_value(); })) {
```

Comment:
when does this check fail?

Reply:
If there is no `bloom_filter_offset` present in any of the row groups, this would exit early.

Reply:
I get that, but we don't enter this function when there are no columns with equality predicate(s). So when do we have the predicates, but no offsets (when file is written without bloom filters)?

Reply:
Yes, exactly!